edb/server/compiler_pool/pool.py

```diff
@@ -209,20 +209,21 @@ async def _attach_worker(self, pid: int, init_args, init_args_pickled):
         await worker._attach(init_args_pickled)
         self._report_worker(worker)
 
-        self._workers.add(worker)
+        self._workers[pid] = worker
         self._workers_queue.release(worker)
-        return worker
-
-    async def start(self):
-        if self._running is not None:
-            raise RuntimeError(
-                'the compiler pool has already been started once')
 
-        self._workers_queue = queue.WorkerQueue(self._loop)
+        logger.debug("Worker with PID %s is ready.", pid)
+        if (
+            not self._ready_evt.is_set()
+            and len(self._workers) == self._pool_size
+        ):
+            logger.info("All %s compiler Workers are ready.", self._pool_size)
+            self._ready_evt.set()
 
-        await self._server.start()
-        self._running = True
+        return worker
 
+    @functools.lru_cache(maxsize=None)
+    def _get_init_args(self):
```

Comment:
Meanwhile, I did find a question regarding the worker's init args. As now we may start workers at any time throughout the server's lifetime, is it okay to keep using the cached init args like `DatabaseState` cached at the time of the server start? Looks like `__sync__` in `worker.py` is used to update the worker state, but I'm not so sure.

Reply:
It's OK to keep using the cached args -- the IO process will send the updated bits with the first request to the worker. We can also update the args if necessary (and cache the new version) but I'm not sure it's worth the work.